app/src/main/cpp/skyline/services/nvdrv/driver.cpp

```diff
@@ -1,61 +1,146 @@
-    Driver::Driver(const DeviceState &state) : state(state), hostSyncpoint(state) {}
-
-    u32 Driver::OpenDevice(std::string_view path) {
-        state.logger->Debug("Opening NVDRV device ({}): {}", fdIndex, path);
-
-        switch (util::Hash(path)) {
-            #define NVDEVICE(type, name, devicePath)                     \
-                case util::Hash(devicePath): {                           \
-                    std::shared_ptr<device::type> device{};              \
-                    if (name.expired()) {                                \
-                        device = std::make_shared<device::type>(state);  \
-                        name = device;                                   \
-                    } else {                                             \
-                        device = name.lock();                            \
-                    }                                                    \
-                    devices.push_back(device);                           \
-                    break;                                               \
-                }
-            NVDEVICE_LIST
-            #undef NVDEVICE
+    Driver::Driver(const DeviceState &state) : state(state), core(state) {}
+
+    NvResult Driver::OpenDevice(std::string_view path, FileDescriptor fd, const SessionContext &ctx) {
+        state.logger->Debug("Opening NvDrv device ({}): {}", fd, path);
+        auto pathHash{util::Hash(path)};
+
+        #define DEVICE_SWITCH(cases) \
+            switch (pathHash) {      \
+                cases;               \
+                default:             \
+                    break;           \
+            }
+
+        #define DEVICE_CASE(path, object) \
+            case util::Hash(path): \
+                devices.emplace(fd, std::make_unique<device::object>(state, core, ctx)); \
+                return NvResult::Success;
+
+        DEVICE_SWITCH(
+            DEVICE_CASE("/dev/nvmap", NvMap)
+            DEVICE_CASE("/dev/nvhost-ctrl", nvhost::Ctrl)
+        );
+
+        if (ctx.perms.AccessGpu) {
+            DEVICE_SWITCH(
+                DEVICE_CASE("/dev/nvhost-as-gpu", nvhost::AsGpu)
+                DEVICE_CASE("/dev/nvhost-ctrl-gpu", nvhost::CtrlGpu)
+                DEVICE_CASE("/dev/nvhost-gpu", nvhost::GpuChannel)
+            );
+        }/*
+
+        if (ctx.perms.AccessVic) {
+            switch (pathHash) {
+                ENTRY("/dev/nvhost-vic", nvhost::Vic)
+                default:
+                    break;
+            }
+        }
 
+        if (ctx.perms.AccessVideoDecoder) {
+            switch (pathHash) {
+                ENTRY("/dev/nvhost-nvdec", nvhost::NvDec)
+                default:
+                    break;
+            }
+        }
+
+        if (ctx.perms.AccessJpeg) {
+            switch (pathHash) {
+                ENTRY("/dev/nvhost-nvjpg", nvhost::NvJpg)
+                default:
+                    break;
+            }
+        }*/
```

Comment:
If this is intentional, can you have a proper comment for it and place the `/*` and `*/` better